crates/parser/src/grammar/types.rs

```diff
@@ -43,6 +44,12 @@ pub fn parse_struct_def(
                 functions.push(parse_fn_def(par, pub_qual)?);
             }
             TokenKind::BraceClose => {
+                if pub_qual.is_some() {
+                    let tok: Token = par.next()?;
+                    par.unexpected_token_error(&tok, "failed to parse struct definition", vec![]);
+                    return Err(ParseFailed);
+                }
```

Comment:
```suggestion
            TokenKind::BraceClose if pub_qual.is_none() => {

```

It seems the if block is the same as the block of the `_` arm, so I think it would be better to use the match guard or return a more detailed error message if the `TokenKind::BraceClose` has its own check for `pub_qual`.

Reply:
Done! decided to go with the former one.